### PR TITLE
SNOW-1256908 Implement structured objects

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description

SNOW-1256908 Implementation of structured objects

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: converter.go [hunk 1]
```diff
@@ -631,6 +594,184 @@ func arrowToValue(
...
+			return f
+		}
+		return fmt.Sprintf("%.*f", scale, float64(val)/math.Pow10(scale))
+	}
+	return nil
+}
+
+func arrowInt16ToValue(srcValue *array.Int16, rowIdx int, higherPrecision bool, scale int) snowflakeValue {
+	if !srcValue.IsNull(rowIdx) {
+		val := srcValue.Value(rowIdx)
```
Comment (sfc-gh-dheyman-1) [inline converter.go:699]: I see that there is some duplication among arrowInt**ToValue methods. Perhaps extract/generify this duplicated logic?

File: datetime.go [hunk 2]
```diff
@@ -60,3 +60,34 @@ func snowflakeFormatToGoFormat(sfFormat string) (string, error) {
 	}
 	return res, nil
 }
+
+func dateTimeFormatByType(dateTimeType string, params map[string]*string) (string, error) {
```
Comment (sfc-gh-dheyman-1) [inline datetime.go:64]: test for that?

File: converter.go [hunk 3]
```diff
@@ -631,6 +594,184 @@ func arrowToValue(
...
+
+func arrowTimeToValue(srcValue arrow.Array, rowIdx int, scale int) snowflakeValue {
+	t0 := time.Time{}
+	if srcValue.DataType().ID() == arrow.INT64 {
+		if !srcValue.IsNull(rowIdx) {
+			return t0.Add(time.Duration(srcValue.(*array.Int64).Value(rowIdx) * int64(math.Pow10(9-scale))))
+		}
+	} else {
+		if !srcValue.IsNull(rowIdx) {
+			return t0.Add(time.Duration(int64(srcValue.(*array.Int32).Value(rowIdx)) * int64(math.Pow10(9-scale))))
```
Comment (sfc-gh-dheyman-1) [inline converter.go:769]: don't we have a util for this math logic?

File: query.go [hunk 4]
```diff
@@ -47,13 +47,21 @@ type contextData struct {
...
+	Precision  int64           `json:"precision"`
+	Scale      int64           `json:"scale"`
+	Nullable   bool            `json:"nullable"`
+}
+
+type fieldMetadata struct {
+	Name   string          `json:"name"`
+	Type   string          `json:"type"`
+	Scale  int             `json:"scale"`
+	Fields []fieldMetadata `json:"fields"`
```
Comment (sfc-gh-pmotacki) [inline query.go:64]: Here should be only 4 properties

File: datetime.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply datetime.go:64]: It is tested via integration test. Do you see a value in unit testing?

File: query.go (hunk 4 above)
Comment (sfc-gh-pfus) [reply query.go:64]: Yes, this is enough - we don't need remaining fields.

**Review by sfc-gh-pmotacki — APPROVED** [review]
LGTM

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

**Pull request merged**